### PR TITLE
Add type hints to `Operator`, `Operation`, `Observable` and Operation classes in `ops/qubit`

### PR DESCRIPTION
**Pull request by Shiro-Raven**
**Context:**
Part of the ongoing effort for better type hints.

**Description of the Change:**
- Type hints have been added.
- New type `WireTypes = Union[qml.Wires, Iterable[Hashable]]` has been added to `qml.typing`.
- New type `FlatPytree = tuple[Iterable[Any], Hashable]` has been added to `qml.operation`

[[sc-66106](https://app.shortcut.com/xanaduai/story/66106)]

### CODE REVIEW

File: pennylane/pauli/grouping/group_observables.py [hunk 1]
```diff
@@ -176,7 +183,12 @@ def colour_pauli_graph(self):
         return self.grouped_paulis
 
 
-def group_observables(observables, coefficients=None, grouping_type="qwc", method="rlf"):
+def group_observables(
+    observables: list["qml.Observable"],
```
Comment (mudit2812) [inline pennylane/pauli/grouping/group_observables.py:187]:
```suggestion
    observables: list["qml.operation.Observable"],
```

Comment (mudit2812) [reply pennylane/pauli/grouping/group_observables.py:187]: Same change anywhere else it might have been added.

File: pennylane/operation.py [hunk 2]
```diff
@@ -406,6 +407,9 @@ def _mod_and_round(x, mod_val):
     return str([id(d) if qml.math.is_abstract(d) else _mod_and_round(d, mod_val) for d in op.data])
 
 
+FlattenOutput = tuple[tuple[TensorLike, ...], tuple[Wires, tuple[tuple[str, Any], ...]]]
```
Comment (mudit2812) [inline pennylane/operation.py:410]: I wasn't originally able to tell what this type means. Maybe this instead?
```suggestion
FlatPytree = tuple[tuple[TensorLike, ...], tuple[Wires, tuple[tuple[str, Any], ...]]]
```

**Review by mudit2812 — COMMENTED** [review]
Looks good, I just left a few comments. I saw some instances of `qml.Operator`, `qml.Observable`, `qml.Operation`, which need to be updated to `qml.operation.Operator`, and so on. I also left a few suggestions for the naming of the custom type hints. Let me know what you think.

**Review by mudit2812 — APPROVED** [review]

File: pennylane/operation.py [hunk 3]
```diff
@@ -406,6 +407,9 @@ def _mod_and_round(x, mod_val):
     return str([id(d) if qml.math.is_abstract(d) else _mod_and_round(d, mod_val) for d in op.data])
 
 
+FlatPytree = tuple[tuple[TensorLike, ...], tuple[Wires, tuple[tuple[str, Any], ...]]]
```
Comment (albi3ro) [inline pennylane/operation.py:410]:
```suggestion
FlatPytree = tuple[tuple[TensorLike, ...], Hashable]
```

Happens to take that particular form as default behavior in the base class, but the actual interface just requires that it be hashable.  Maybe we should update the docstring to reflect that as well?

File: pennylane/operation.py [hunk 4]
```diff
@@ -958,24 +964,29 @@ def terms(self):  # pylint: disable=no-self-use
         """
         raise TermsUndefinedError
 
-    num_wires = AnyWires
+    num_wires: WiresEnum = AnyWires
```
Comment (albi3ro) [inline pennylane/operation.py:967]:
```suggestion
    num_wires: Union[int, WiresEnum] = AnyWires
```

File: pennylane/operation.py [hunk 5]
```diff
@@ -958,24 +964,29 @@ def terms(self):  # pylint: disable=no-self-use
...
 
     @property
-    def id(self):
+    def id(self) -> str:
         """Custom string to label a specific operator instance."""
         return self._id
 
     @name.setter
-    def name(self, value):
+    def name(self, value: Any):
```
Comment (albi3ro) [inline pennylane/operation.py:981]:
```suggestion
    def name(self, value: str):
```

File: pennylane/operation.py [hunk 6]
```diff
@@ -1551,7 +1573,7 @@ def simplify(self) -> "Operator":  # pylint: disable=unused-argument
         """
         return self
 
-    def __add__(self, other):
+    def __add__(self, other: Union["Operator", TensorLike]):
```
Comment (albi3ro) [inline pennylane/operation.py:1576]:
```suggestion
    def __add__(self, other: Union["Operator", TensorLike]) -> "Operator":
```

File: pennylane/operation.py [hunk 7]
```diff
@@ -1567,49 +1589,49 @@ def __add__(self, other):
 
     __radd__ = __add__
 
-    def __mul__(self, other):
+    def __mul__(self, other: Union[Callable, TensorLike]):
```
Comment (albi3ro) [inline pennylane/operation.py:1592]:
```suggestion
    def __mul__(self, other: Union[Callable, TensorLike]) -> "Operator":
```

File: pennylane/operation.py [hunk 8]
```diff
@@ -1567,49 +1589,49 @@ def __add__(self, other):
...
+    def __truediv__(self, other: TensorLike):
         """The division between an Operator and a number."""
         if isinstance(other, TensorLike):
             return self.__mul__(1 / other)
         return NotImplemented
 
     __rmul__ = __mul__
 
-    def __matmul__(self, other):
+    def __matmul__(self, other: "Operator"):
```
Comment (albi3ro) [inline pennylane/operation.py:1608]:
```suggestion
    def __matmul__(self, other: "Operator") -> "Operator":
```

File: pennylane/operation.py [hunk 9]
```diff
@@ -1647,7 +1669,9 @@ def _flatten(self):
         return self.data, (self.wires, hashable_hyperparameters)
 
     @classmethod
-    def _unflatten(cls, data, metadata):
+    def _unflatten(
+        cls, data: tuple[TensorLike, ...], metadata: tuple[Wires, tuple[tuple[str, Any], ...]]
```
Comment (albi3ro) [inline pennylane/operation.py:1673]:
```suggestion
        cls, data: tuple[Any], metadata: hashable
```
I'm not quite sure how to type hint "terminal leaf or other pytree".  So maybe just "Any" is good enough...

Also, the `metadata` can be any type of hashable information.

File: pennylane/operation.py [hunk 10]
```diff
@@ -1950,11 +1979,13 @@ def _queue_category(self):
...
 
     @property
-    def is_hermitian(self):
+    def is_hermitian(self) -> bool:
         """All observables must be hermitian"""
         return True
 
-    def __matmul__(self, other):
+    def __matmul__(
+        self, other: Union["Operator", "Tensor", "qml.ops.Hamiltonian", "qml.ops.LinearCombination"]
```
Comment (albi3ro) [inline pennylane/operation.py:1987]: Given they are all instances of type `Operator`, do we need to specify different types of special operator?

File: pennylane/ops/qubit/arithmetic_ops.py [hunk 11]
```diff
@@ -152,7 +154,7 @@ def compute_matrix():  # pylint: disable=arguments-differ
         )
 
     @staticmethod
-    def compute_decomposition(wires):
+    def compute_decomposition(wires: WiresLike) -> list["qml.operation.Operator"]:
```
[the remark this answers is not in the document]
Comment (Shiro-Raven) [reply pennylane/ops/qubit/arithmetic_ops.py:157]: Nope you are right

File: pennylane/operation.py [hunk 12]
```diff
@@ -2054,13 +2084,14 @@ def __mul__(self, a):
 
     __rmul__ = __mul__
 
-    def __sub__(self, other):
+    def __sub__(self, other: "Observable") -> "Observable":
```
Comment (albi3ro) [inline pennylane/operation.py:2087]:
```suggestion
    def __sub__(self, other: Operator) -> Operator:
```

We can probably go back and update the docstring at some point.

File: pennylane/ops/qubit/non_parametric_ops.py [hunk 13]
```diff
@@ -80,11 +87,11 @@ def compute_matrix():  # pylint: disable=arguments-differ
 
     @staticmethod
     @lru_cache()
-    def compute_sparse_matrix():  # pylint: disable=arguments-differ
+    def compute_sparse_matrix() -> sparse.csc_matrix:  # pylint: disable=arguments-differ
```
Comment (albi3ro) [inline pennylane/ops/qubit/non_parametric_ops.py:90]:
```suggestion
    def compute_sparse_matrix() -> sparse.csr_matrix:  # pylint: disable=arguments-differ
```

File: pennylane/ops/qubit/non_parametric_ops.py [hunk 14]
```diff
@@ -109,7 +116,7 @@ def compute_eigvals():  # pylint: disable=arguments-differ
         return pauli_eigs(1)
 
     @staticmethod
-    def compute_diagonalizing_gates(wires):
+    def compute_diagonalizing_gates(wires: WiresLike) -> list["qml.operation.Operator"]:
```
Comment (albi3ro) [inline pennylane/ops/qubit/non_parametric_ops.py:119]:
```suggestion
    def compute_diagonalizing_gates(wires: WiresLike) -> list[qml.operation.Operator]:
```

File: pennylane/ops/qubit/non_parametric_ops.py [hunk 15]
```diff
@@ -328,19 +340,19 @@ def compute_decomposition(wires):
             qml.PhaseShift(np.pi / 2, wires=wires),
         ]
 
-    def adjoint(self):
+    def adjoint(self) -> "PauliX":
         return X(wires=self.wires)
 
-    def pow(self, z):
+    def pow(self, z: Union[int, float]) -> list["qml.operation.Operator"]:
```
Comment (albi3ro) [inline pennylane/ops/qubit/non_parametric_ops.py:346]:
```suggestion
    def pow(self, z: Union[int, float]) -> list[qml.operation.Operator]:
```

File: pennylane/ops/qubit/non_parametric_ops.py [hunk 16]
```diff
@@ -391,27 +403,32 @@ class PauliY(Observable, Operation):
 
     _queue_category = "_ops"
 
-    def __init__(self, wires=None, id=None):
+    def __init__(self, wires: Optional[WiresLike] = None, id: Optional[str] = None):
```
Comment (albi3ro) [inline pennylane/ops/qubit/non_parametric_ops.py:406]:
```suggestion
    def __init__(self, wires: WiresLike, id: Optional[str] = None):
```
Optional while we're here?

File: pennylane/ops/qubit/non_parametric_ops.py [hunk 17]
```diff
@@ -461,7 +478,7 @@ def compute_eigvals():  # pylint: disable=arguments-differ
         return pauli_eigs(1)
 
     @staticmethod
-    def compute_diagonalizing_gates(wires):
+    def compute_diagonalizing_gates(wires: WiresLike) -> list["qml.operation.Operator"]:
```
Comment (albi3ro) [inline pennylane/ops/qubit/non_parametric_ops.py:481]:
```suggestion
    def compute_diagonalizing_gates(wires: WiresLike) -> list[qml.operation.Operator]:
```

File: pennylane/ops/qubit/non_parametric_ops.py [hunk 18]
```diff
@@ -490,7 +507,7 @@ def compute_diagonalizing_gates(wires):
         ]
 
     @staticmethod
-    def compute_decomposition(wires):
+    def compute_decomposition(wires: WiresLike) -> list["qml.operation.Operator"]:
```
Comment (albi3ro) [inline pennylane/ops/qubit/non_parametric_ops.py:510]:
```suggestion
    def compute_decomposition(wires: WiresLike) -> list[qml.operation.Operator]:
```

File: pennylane/ops/qubit/non_parametric_ops.py [hunk 19]
```diff
@@ -517,16 +534,16 @@ def compute_decomposition(wires):
             qml.PhaseShift(np.pi / 2, wires=wires),
         ]
 
-    def adjoint(self):
+    def adjoint(self) -> "PauliY":
         return Y(wires=self.wires)
 
-    def pow(self, z):
+    def pow(self, z: Union[float, int]) -> list["qml.operation.Operator"]:
```
Comment (albi3ro) [inline pennylane/ops/qubit/non_parametric_ops.py:540]:
```suggestion
    def pow(self, z: Union[float, int]) -> list[qml.operation.Operator]:
```

File: pennylane/ops/qubit/non_parametric_ops.py [hunk 20]
```diff
@@ -575,27 +592,32 @@ class PauliZ(Observable, Operation):
 
     _queue_category = "_ops"
 
-    def __init__(self, wires=None, id=None):
+    def __init__(self, wires: Optional[WiresLike] = None, id: Optional[str] = None):
```
Comment (albi3ro) [inline pennylane/ops/qubit/non_parametric_ops.py:595]:
```suggestion
    def __init__(self, wires: WiresLike, id: Optional[str] = None):
```

File: pennylane/ops/qubit/non_parametric_ops.py [hunk 21]
```diff
@@ -645,7 +667,9 @@ def compute_eigvals():  # pylint: disable=arguments-differ
         return pauli_eigs(1)
 
     @staticmethod
-    def compute_diagonalizing_gates(wires):  # pylint: disable=unused-argument
+    def compute_diagonalizing_gates(  # pylint: disable=unused-argument
+        wires: WiresLike,
+    ) -> list["qml.operation.Operator"]:
```
Comment (albi3ro) [inline pennylane/ops/qubit/non_parametric_ops.py:672]:
```suggestion
    ) -> list[qml.operation.Operator]:
```

File: pennylane/ops/qubit/non_parametric_ops.py [hunk 22]
```diff
@@ -671,7 +695,7 @@ def compute_diagonalizing_gates(wires):  # pylint: disable=unused-argument
         return []
 
     @staticmethod
-    def compute_decomposition(wires):
+    def compute_decomposition(wires: WiresLike) -> list["qml.operation.Operator"]:
```
Comment (albi3ro) [inline pennylane/ops/qubit/non_parametric_ops.py:698]:
```suggestion
    def compute_decomposition(wires: WiresLike) -> list[qml.operation.Operator]:
```

File: pennylane/ops/qubit/non_parametric_ops.py [hunk 23]
```diff
@@ -807,7 +831,7 @@ def compute_eigvals():  # pylint: disable=arguments-differ
         return np.array([1, 1j])
 
     @staticmethod
-    def compute_decomposition(wires):
+    def compute_decomposition(wires: WiresLike) -> np.ndarray:
```
Comment (albi3ro) [inline pennylane/ops/qubit/non_parametric_ops.py:834]:
```suggestion
    def compute_decomposition(wires: WiresLike) -> List[qml.operation.Operator]:
```

File: pennylane/ops/qubit/non_parametric_ops.py [hunk 24]
```diff
@@ -829,7 +853,7 @@ def compute_decomposition(wires):
         """
         return [qml.PhaseShift(np.pi / 2, wires=wires)]
 
-    def pow(self, z):
+    def pow(self, z: Union[int, float]) -> list["qml.operation.Operator"]:
```
Comment (albi3ro) [inline pennylane/ops/qubit/non_parametric_ops.py:856]:
```suggestion
    def pow(self, z: Union[int, float]) -> list[qml.operation.Operator]:
```

File: pennylane/ops/qubit/observables.py [hunk 25]
```diff
@@ -292,7 +301,7 @@ class SparseHamiltonian(Observable):
 
     grad_method = None
 
-    def __init__(self, H, wires=None, id=None):
+    def __init__(self, H: csr_matrix, wires: Optional[WiresLike] = None, id: Optional[str] = None):
```
Comment (albi3ro) [inline pennylane/ops/qubit/observables.py:304]:
```suggestion
    def __init__(self, H: csr_matrix, wires: WiresLike, id: Optional[str] = None):
```

File: pennylane/ops/qubit/parametric_ops_multi_qubit.py [hunk 26]
```diff
@@ -69,16 +71,20 @@ class MultiRZ(Operation):
     grad_method = "A"
     parameter_frequencies = [(1,)]
 
-    def _flatten(self):
+    def _flatten(self) -> FlatPytree:
         return self.data, (self.wires, tuple())
 
-    def __init__(self, theta, wires=None, id=None):
+    def __init__(
+        self, theta: TensorLike, wires: Optional[WiresLike] = None, id: Optional[str] = None
```
Comment (albi3ro) [inline pennylane/ops/qubit/parametric_ops_multi_qubit.py:78]:
```suggestion
        self, theta: TensorLike, wires: WiresLike, id: Optional[str] = None
```

File: pennylane/ops/qubit/parametric_ops_multi_qubit.py [hunk 27]
```diff
@@ -265,7 +273,13 @@ class PauliRot(Operation):
     def _primitive_bind_call(cls, theta, pauli_word, wires=None, id=None):
         return super()._primitive_bind_call(theta, pauli_word=pauli_word, wires=wires, id=id)
 
-    def __init__(self, theta, pauli_word, wires=None, id=None):
+    def __init__(
+        self,
+        theta: TensorLike,
+        pauli_word: Literal["I", "X", "Y", "Z"],
```
Comment (albi3ro) [inline pennylane/ops/qubit/parametric_ops_multi_qubit.py:279]: So strings like `"XYZ"` are also valid.   I don't think we are going to be able to get more specific than `str`. Any string *containing* only XYZI is allowed.
```suggestion
        pauli_word: str,
```

File: pennylane/ops/qubit/parametric_ops_multi_qubit.py [hunk 28]
```diff
@@ -265,7 +273,13 @@ class PauliRot(Operation):
...
+        wires: Optional[WiresLike] = None,
```
Comment (albi3ro) [inline pennylane/ops/qubit/parametric_ops_multi_qubit.py:280]:
```suggestion
        wires: WiresLike,
```

File: pennylane/ops/qubit/parametric_ops_multi_qubit.py [hunk 29]
```diff
@@ -588,7 +613,7 @@ def __init__(self, phi, dim, wires, id=None):
         self.hyperparameters["dimension"] = (dim, 2 ** len(wires))
 
     @staticmethod
-    def compute_matrix(*params, **hyperparams):
+    def compute_matrix(*params: TensorLike, **hyperparams) -> TensorLike:
         """Get the matrix representation of Pi-controlled phase unitary."""
         phi = params[0]
         d, t = hyperparams["dimension"]
```
Comment (albi3ro) [inline pennylane/ops/qubit/parametric_ops_multi_qubit.py:619]:
```suggestion
    def compute_matrix(phi: TensorLike, dimension: tuple[int, int]) -> TensorLike:
        """Get the matrix representation of Pi-controlled phase unitary."""
        d, t = dimension
```

File: pennylane/ops/qubit/special_unitary.py [hunk 30]
```diff
@@ -705,7 +710,11 @@ class TmpPauliRot(PauliRot):
     has_matrix = False
 
     @staticmethod
-    def compute_decomposition(theta, wires, pauli_word):
+    def compute_decomposition(
+        theta: TensorLike,
+        wires: WiresLike,
+        pauli_word: Literal["I", "X", "Y", "Z"],
```
Comment (albi3ro) [inline pennylane/ops/qubit/special_unitary.py:716]:
```suggestion
        pauli_word: str,
```

File: pennylane/pauli/grouping/group_observables.py [hunk 31]
```diff
@@ -176,7 +183,12 @@ def colour_pauli_graph(self):
         return self.grouped_paulis
 
 
-def group_observables(observables, coefficients=None, grouping_type="qwc", method="rlf"):
+def group_observables(
+    observables: list["qml.operation.Observable"],
```
Comment (albi3ro) [inline pennylane/pauli/grouping/group_observables.py:187]:
```suggestion
    observables: list["qml.operation.Operator"],
```

New opmath objects no longer inherit from `Observable`.

File: pennylane/wires.py [hunk 32]
```diff
@@ -695,5 +696,7 @@ def __rxor__(self, other):
         return Wires((set(_process(other)) ^ set(self.labels)))
 
 
+WiresLike = Union[Wires, Iterable[Hashable]]
```
Comment (albi3ro) [inline pennylane/wires.py:699]:
```suggestion
WiresLike = Union[Wires, Iterable[Hashable], Hashable]
```
For example, we will accept `wires=0`.  0 is an individual wire label, not an iterable of wire labels.

**Review by albi3ro — APPROVED** [review]
🎉

**Pull request merged**